File: plugins/harness/skills/harness-install/assets/bun/runtime/rules/require-template-groups.ts
#!/usr/bin/env bun
import type { Finding, HarnessCheckRule, HarnessManifest, RuleContext } from "../harness-check-rule";

/**
 * Require specified template groups to exist.
 */
export const requireTemplateGroupsRule = (ctx: RuleContext): HarnessCheckRule => ({
  applies(manifest: HarnessManifest): boolean {
    const section = manifest.requireTemplateGroups;
    if (typeof section !== "object" || section === null) {
      return false;
    }
    const enabled = (section as { enabled?: unknown }).enabled;
    if (enabled === false) {
      return false;
    }
    const entry = ctx.readJsonObject((section as Record<string, unknown>).parameters);
    return ctx.readStringArray(entry.groups).length > 0;
  }

  validate(_root: string, manifest: HarnessManifest): readonly Finding[] {
    const section = ctx.readJsonObject(manifest.requireTemplateGroups);
    const parameters = ctx.readJsonObject(section.parameters);
    const targetRoot = typeof parameters.targetRoot === "string" ? parameters.targetRoot : "";
    const groups = ctx.readStringArray(parameters.groups);
    return groups.flatMap((group) => {
      const path = `${targetRoot}/${group}`;
      return ctx.isDirectory(path)
        ? []
        : [
            {
              severity: ctx.severityOf(manifest, "requireTemplateGroups"),
              category: "requireTemplateGroups",
              message: `missing template group: ${path}`,
            },
          ];
    });
  }

});

File: plugins/harness/skills/harness-install/assets/uv/runtime/rules/require_template_groups.py
#!/usr/bin/env -S uv run
# -*- coding: utf-8 -*-
"""Require template groups rule."""

from pathlib import Path

from harness_check_rule import Finding, HarnessCheckRule

from ._utils import is_safe_directory, severity_for


class RequireTemplateGroupsRule(HarnessCheckRule):
    """Validate requireTemplateGroups check."""

    category = "requireTemplateGroups"

    def applies(self, manifest: dict) -> bool:
        """Check if this rule applies to the manifest."""
        section = manifest.get(self.category)
        if not isinstance(section, dict):
            return False
        enabled = section.get("enabled", True)
        return enabled is not False

    def validate(self, root: Path, manifest: dict) -> list[Finding]:
        """Validate requireTemplateGroups check."""
        section = manifest.get(self.category, {})
        if not isinstance(section, dict):
            return []
        params = section.get("parameters", {})
        if not isinstance(params, dict):
            return []
        target_root = params.get("targetRoot", "docs/harness/templates")
        groups = params.get("groups", [])
        if not isinstance(groups, list):
            return []
        messages = section.get("messages", {})
        if not isinstance(messages, dict):
            return []
        template = messages.get("default", "missing template group: {targetRoot}/{group}")
        return [
            Finding(
                severity_for(manifest, self.category),
                self.category,
                template.format(targetRoot=target_root, group=group),
            )
            for group in groups
            if isinstance(group, str) and not is_safe_directory(root / target_root / group)
        ]
